### PR TITLE
 Issues 61 - Reducing silence threshold to ignore a sentence

### PR DESCRIPTION
**Pull request by jdorleans**
This PR solves #61.
It turns out the silence threshold of 1 second was big enough to ignore many sentences causing the "Go ahead behaviour" instead of processing the query. We have decided to reduce it to 0.5 seconds.

### CODE REVIEW

**Comment by aatchison** [comment]
I still get the go-ahead behavior when I start the request with "Hey Mycroft."
I get a multiple intent parsing error or WA search because it splits: "Hey Mycroft, what time is it?" into: 'haim','time is it.'  Every once in a while it succeeds.

Otherwise:
"Mycroft, what time is it."
"What time is it, Mycroft."
"Hey, Mycroft". "Go ahead." What time is it?"
are all working about 7/10 times on my cruddy laptop mic.

**Comment by clusterfudge** [comment]
are you suggesting that "what time is it" is an utterance < 1s? That seems unnlikely to me.

**Comment by ryanleesipes** [comment]
@clusterfudge, that is exactly what I'm suggesting. Before the threshold was moved to .5, "What time is it, Mycroft?" And "What's the weather, Mycroft?" were failing.

**Comment by jdorleans** [comment]
Ok guys, we have to be careful here and focus on the scope of this task. After many trails we have noticed that reducing the threshold is beneficial and it prevents short sentences to be ignored like: **what's the time**, **what's the weather** or simply *stop* (from Mycroft Stop), *record* etc.

**Comment by Wolfgange3311999** [comment]
What's the length of the utterance "stop" when detected with a high quality mic? Maybe it should be reduced even further.

However, with low quality mics (in my experience), even a short "stop" can take 0.8-1.5 seconds. Perhaps this should be a preference (that ideally gets calibrated)?

**Comment by aatchison** [comment]
I think we need to think about how we might test different values? Could we detect mic quality in regards to noise or range of the recording? Pretty vague I know.

**Comment by Wolfgange3311999** [comment]
Just brainstorming, to calibrate both this value and maybe some others such as the dynamic adjustment ratio in the speech recognition library, after pairing Mycroft could optionally perform a calibration session where if on a computer you set the volume to the maximum in a quiet environment and it outputs a beep of 1 second measuring the extra length after the beep from whatever it is that causes the variance between mics.

However, for the scope of this PR I'd say someone should test `Mycroft stop` a few times on a good mic and find the maximum acceptable silence threshold.

**Comment by ryanleesipes** [comment]
As @Wolfgange3311999 points out, "Mycroft stop" does not work well with this threshold. Other than that this appears to work quite well.

**Comment by clusterfudge** [comment]
The original purpose of this threshold was twofold, one reason no longer applying. 

1) Decrease the number of calls to cerberus, to minimize unnecessary API calls on a paid resource, or decrease load on a service we own (in the future of openSTT)
2) To decrease the number of STT calls being made on the recognizer loop, which dramatically increased interaction latency.

Since the STT calls have been parallelized, 2 is no longer a problem. If we're willing to just eat the cost of the extra API calls, we may as well drop the threshold to 0 (or remove the threshold code altogether). I haven't seen any data in this ticket yet as to what makes a good threshold, and I don't want to keep guessing with unexpected results.

**Comment by jdorleans** [comment]
The threshold is still valid. Every time we remove "Mycroft" wake word from a sentence (using the word extractor algorithm), it will generate 2 sentences and one of them is very likely to be only silence. For instance, the utterance: "Mycroft, what time is it?" will result in the following 2 sentences: 
- S1: "" (silence, mycroft was removed)
- S2: "what time is it"
We do not want to process S1. Then, the silence threshold will helps us to ignore it.

Our new Word Extractor adds less silence to the split sentences than before. Which means, we need to recalibrate the threshold. 

I'll do some new tests with "Mycroft stop", because it seems to be the shortest query we have. Finally, I'll adjust the threshold again and provide a new commit.

**Comment by clusterfudge** [comment]
In your last comment you say "We do not want to process S1". Can you elaborate on which processing we want to avoid? Server-side STT? Local intent determination? Something else?

**Comment by jdorleans** [comment]
We wanna discard S1, since it's only silence and it is not useful for further processing both local and server-side.

**Comment by clusterfudge** [comment]
So, realistically, either S1 or S2 can be the incorrect phrase (this is currently broken and referenced in #61 ). The original problem was that sometimes S1 **or** S2 would not be useful, **and** that negatively impacted user experience (via latency in the main speech loop).

The recognizer as is does not suffer from the previously mentioned latency problems because of parallel STT calls. The behavior of the speech loop is identical whether we drop the audio locally or get back an error/empty response from STT, so there's no change to user experience here. 

My main argument is that we have a magic number that we're changing because it "works better", without any evidence in either direction. My understanding is that we can remove this code and have identical user behavior.  I am a huge fan of gaining correctness (thus improving user experience) by removing complexity over tinkering with magic numbers or introducing more complexity. 

TL/DR; This change claims "improvements", but it's an empty claim. Please make your case with unit tests, or consider a change that can be validated and enforced (via automated tests) long term.

**Comment by ryanleesipes** [comment]
Let's get some unit tests in here, and get this in. Mycroft is seemingly much more attentive now, and adjusting the min_audio_size does fix the problem of him responding with "Go Ahead" vs just doing the action.

We need some tests in here that validate that a change in this regard is the best move (and I agree with @clusterfudge - perhaps we should get rid of this "magic number" altogether). Just let me know what we need, and we'll get it (audio files or whatever they are).

**Comment by clusterfudge** [comment]
@ryanleesipes you appear to be trying to sneak some StopSkill changes into this PR. Please don't.

**Comment by ryanleesipes** [comment]
Yeah, that was a mistake (literally clicked the wrong thing). Needs to be reverted.

**Comment by ethanaward** [comment]
Seems to be working well for me.

**Comment by ryanleesipes** [comment]
Tested, LGTM, Merge.
